### ISSUE TITLE
Resolve ampersand to component-specific class, not instance-specific class

### ISSUE THREAD

**Issue by jsamr** [styled-components/styled-components#1275]
[In the bellow example](https://www.webpackbin.com/bins/-KxdMbHX3HjjWFzAY_bv), one would expect the `margin-left` rule to apply to the button themed with the admin palette. But this is not the case : the ampersand is resolved to the instance class. 
``` javascript
import React from 'react';
import styled, { ThemeProvider } from 'styled-components';

const normalPalette = {
   background: 'black',
   foreground: 'white'
}

const adminPalette = {
   background: 'white',
   foreground: 'black'
}

// Create a <Button> react component which background and foreground rules
// are theme-dependant, but the sibling rule with left margin should be transversal
// Unfortunately, it doesn't work because the two sibling button instances have
// different class names.
const Button = styled.button`
  ${({ theme }) => `
    background-color: ${theme.background};
    color: ${theme.foreground};
  `}
  &+& {
    margin-left: 10px;
  }
`;

export default () => {
  return (
    <div>
      <ThemeProvider theme={normalPalette}>
        <Button> normal button </Button>
      </ThemeProvider>
      <ThemeProvider theme={adminPalette}>
        <Button> admin button </Button>
      </ThemeProvider>
    </div>
  );
}
```
The workaround now is to use a direct reference to the component, [as in such example](https://www.webpackbin.com/bins/-KxdNnta1t7Ygpi90EOG). However, this is counter-intuitive, cumbersome and ugly !
Plus, in my use-case, I declare components and their styles in two steps since I am implementing a rich featured theme engine. In such circumstances, I cannot refer to react components classes.

## Feature request (breaking change)

`&` should now be resolved to the component static class selector.

## Alternative (non-breaking)

Offer a specific pseudo class that will be resolved to the static class selector, such as `:sc`.

**Comment by mxstbr** [comment]
This would be a breaking change, but I think it makes total sense to resolve `&` to the component-specific class now that we have it rather than the instance-class.

/cc @philpl @geelen

**Comment by philpl** [comment]
We already had an issue here regarding this but I think we mistakenly closed it: https://github.com/styled-components/styled-components/issues/774

In v3 we might have more tools to solve this, since we'll be able to reliable detect this selector and resolve it differently.

**Comment by sveinburne** [comment]
> Instead all references with two styled components (& + &, & > &, & &) should resolve the second parent selector to the component specific class.

@philpl so that would also include compositions like `&__icon + &__icon`, right ?

What do you thing of the possibility for your library consumer to explicitly chose which class to refer to ? I though of specific namespaced pseudo classes but this is kind of hacky.

**Comment by philpl** [comment]
@sveinburne I don’t think that that’s a good idea.

Resolving the parent selector to the component class only will always lead to unexpected behaviour, so given the choice that will throw people off.

Having a special selector would be a non-standard CSS feature which we don’t want to introduce.

And last but not least, the “${() => Component}” interpolation will continue to work, which eliminates any need for a separate API, really.

Since these selectors that use the parent selector to refer to other components of the same type, but not necessarily the same props, are a special edge case, and I have not seen any other case where workarounds for selectors are necessary.

Thus if we detect these special cases and resolve them differently there’s:

- no new APIs
- no new edge cases
- no need for awareness on the user’s side that this is happening

This detection should become quite trivial in v3, if we pull off ISTF, which I’m working on right now: https://github.com/cssinjs/istf-spec

**Comment by philpl** [comment]
@thysultan how feasible do you think is it to implement this now using a stylis middleware that targets selectors?

If we could detect something like `/(.*&.*)?[+>](.*(&).*)+{/g` and replace the ampersand (parent selector) after `+` or `>` with the component class using a simple regex, would that hurt perf a lot? I think you are a better judge of that than I am

**Comment by thysultan** [comment]
@philpl It is feasible.

As an alternative proposal(Possibly a non-breaking change)...

Given the output s-c emits.

```html
<button class="sc-bdVaJa cVcVQf"> normal button </button>
<button class="sc-bdVaJa fNjRsK"> admin button </button>
```
Why not use the `sc-bdVaJa` class to target all button instances for non dynamic values and `cVcVQf` for dynamic values.

In the example

```js
&+& {
    margin-left: 10px;
}
```

would emit to a `sc-bdVaJa` while

```
  ${({ theme }) => `
    background-color: ${theme.background};
    color: ${theme.foreground};
  `}
```

To different styles targeting `cVcVQf` and `fNjRsK`.

From another angle you might be able to extend this branching of dynamic vs static to making use of the `insertRule` API solely for these dynamic branches.

**Comment by philpl** [comment]
@thysultan from a v2 and stylis perspective it's very hard to determine which rules are dynamic and which aren't. Also, I think this would break down if you had the example selector together with a dynamic selector, i.e.:

```js
  ${elementTag} & + & {
     margin-left: 10px;
  }
```

I think it's harder to do and more error prone with more edge cases being created by an instable dynamic v static analysation. The difference is harder to define than it is said, while I can explain the selector change above with one regex, so I'm unsure whether this won't be more unstable.

> From another angle you might be able to extend this branching of dynamic vs static to making use of the insertRule API solely for these dynamic branches.

This sounds like the static extraction optimisation, which will be hard to do even with v3, but I am thinking of this.

It's pretty hard to do these statis v dynamic analysations efficiently, since we either look at the tag's input before the transformation, or at the raw css with resolved interpolations, or at the final CSS from stylis.

> Possibly a non-breaking change

Btw I think if we'd change the behaviour this dramatically, it definitely will be a breaking change for some cases. But if we just cover the cases that don't work correctly anyway, if I'm not mistaken, it shouldn't negatively influence anyone.

We basically just have to cover `& + & ...` and `& > & ...` where content can be around the ampersand symbols and where the first ampersand is optional.

*New Idea:* So actually, since it doesn't matter what happens to the other ampersand, maybe we can look at the selectors only once and transform them before the CSS reaches stylis? This would make such a transformation extremely cheap, performance wise, but more bug-prone I guess.

**Comment by thysultan** [comment]
I'm positive `${elementTag} & + & { margin-left: 10px; }` and other edge cases would work if you introduce a stylis middleware, an overview of the implementation logic:

1. Replace the interpolation with a static hash/placeholder.
2. Use a middleware to remove and extract all css rules that have the aforementioned placeholder indicating a dynamic value.

You then have the delta – which is all the static content. 

If it is not already the case you could probably do all of this at the component declaration level i.e when you call `const Button = styled.div...`  making the static part a constant time `O(1)` per component instead of a per instance operation.

At that point you now have you processed static css and a hashtable of components and their dynamic parts that you can pass to a new instance of Stylis that only handles dynamic css.

**Comment by philpl** [comment]
@thysultan it’s definitely possible, but let’s keep this issue on topic :smile:

I don’t think it’s worth introducing this right now as it’s a lot of work for such a small issue. Also, we have to “eject” dynamic rules that come before static ones as the static ones will have to keep their precedence. (Explained at the end of this https://gist.github.com/philpl/7b8a021f19ead2ec534091976a1ff6dc)

So I’m not convinced that we should start to introduce the static/dynamic logic and string manipulation for replacing interpolations when I’m specifically working on ISTF to make this part safer as well in the future.

I do have some code for the interpolation placeholders in the Babel plugin if there’s no other way. But since there is another way in this case I think we don’t have to go down this road

**Comment by thysultan** [comment]
Sure, i'll open a new issue with my ideas on that.

>replace the parent selectors / & after + or > with the component class using a simple regex`

@philpl The only hurdle i can see with is that middlewares are not privy to the `&` tokens, since at that point it would be the namespace passed to stylis. Though there is the option of adding a middleware pre-selector context that runs before `&` tokens are converted to the namespace to facilitate this.

**Comment by philpl** [comment]
@thysultan 

>  Though there is the option of adding a middleware pre-selector context that runs before & tokens are converted to the namespace to facilitate this.

So it doesn't exist right now?

**Comment by thysultan** [comment]
@philpl  Correct.

**Comment by probablyup** [comment]
This is done in v4
